buffer/src/main/java/io/netty/buffer/ByteBuf.java

```diff
@@ -307,6 +307,9 @@
     /**
      * Returns {@code true} if and only if this buffer is backed by an
      * NIO direct buffer.
+     * Note: An empty buffer is both direct and heap buffer,
+     * because it is always backed by a zero-length byte array and a direct buffer internally.
+     * See {@link EmptyByteBuf#isDirect()}
```

Comment:
@trustin @chrisvest Adding the doc here doesn't feel right as this is an implementation detail. An custom implementation of `ByteBufAllocator` / `ByteBuf` may not do this.

Reply:
Maybe we should just rephrase this to make it "optional".

Something like:

```
* Note: An empty buffer MAY be both direct and heap buffer,     
* because it is MAY be backed by a zero-length byte array and a direct buffer internally.    
* See {@link EmptyByteBuf#isDirect()}
```

Some comment true for all the others

Reply:
I agree with @normanmaurer here.

Reply:
I see two options here:
- Make it a contract, so it's clear that an empty buffer is always both a heap buffer and a direct buffer; or
- Add a section in the Javadoc of `ByteBuf` that states it is possible for a `ByteBuf` to be backed by both a heap buffer and a direct buffer, mentioning `Unpooled.EMPTY_BUFFER` as an example.

Reply:
@trustin I think the second option is the best as we can't really add the contract now as it would be a "breaking change".

Reply:
+1 that the second option is better. I would also suggest to remove the slight implementation detail suggested in https://github.com/netty/netty/pull/11366/files#r646313599 and use a language like:

```
NOTE: {@link EmptyByteBuf#isDirect()} and {@link EmptyByteBuf#isHeap()} are not mutually 
exclusive, a buffer can be both heap and direct.
```

Reply:
@trustin , @normanmaurer , @NiteshKant, @chrisvest  I think the first option is better, if we use the second option, then there will possible be a non-empty buffer with both a byte array and direct buffer, then we may should modify the logic from method like `ByteBufUtil.writeUtf8(AbstractByteBuf, int, int, CharSequence, int, int)` , because it does not cover both array and direct logic like following:
```
        if (PlatformDependent.hasUnsafe()) {
            if (buffer.hasArray()) {
                return unsafeWriteUtf8(buffer.array(), PlatformDependent.byteArrayBaseOffset(),
                                       buffer.arrayOffset() + writerIndex, seq, start, end);
            }
            if (buffer.hasMemoryAddress()) {
                return unsafeWriteUtf8(null, buffer.memoryAddress(), writerIndex, seq, start, end);
            }
        } else {
            if (buffer.hasArray()) {
                return safeArrayWriteUtf8(buffer.array(), buffer.arrayOffset() + writerIndex, seq, start, end);
            }
            if (buffer.isDirect()) {
                assert buffer.nioBufferCount() == 1;
                final ByteBuffer internalDirectBuffer = buffer.internalNioBuffer(writerIndex, reservedBytes);
                final int bufferPosition = internalDirectBuffer.position();
                return safeDirectWriteUtf8(internalDirectBuffer, bufferPosition, seq, start, end);
            }
        }
        return safeWriteUtf8(buffer, writerIndex, seq, start, end);
    }

Reply:
> then there will possible exist a non-empty buffer with both a byte array and direct buffer

Will implementing such a buffer be wrong though? If an empty buffer can have both byte array and a direct buffer is this technically wrong for an non-empty buffer to have both too? As long as both direct and heap memory has same data, the code you have referred here will work, correct? So, the question is does it help us in any way by making the contract stricter?

Reply:
@NiteshKant As you said:

> As long as both direct and heap memory has same data, the code you have referred here will work, correct?

No it will not work. 
Think about this question: 
How to make both direct and heap memory has same data when you call the public util method like 'ByteBufUtil.writeUtf8(ByteBuf, CharSequence)'? 
The answer is you have to modify the logic of 'ByteBufUtil.writeUtf8(ByteBuf, CharSequence)', right? That is what I said "not good" because it break the origin contract "**if and only if**" in a larger scale.